lib/Sema/CSFix.cpp

```diff
@@ -2151,3 +2152,162 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
+
+    Action walkToTypePre(Type componentTy) override {
+      // In cases where result references a member type, we need to check
+      // whether such type would is resolved to concrete or not.
+      if (auto *member = componentTy->getAs<DependentMemberType>()) {
+        auto memberBaseTy = getBaseTypeOfDependentMemberChain(member);
+
+        auto typeVar = memberBaseTy->getAs<TypeVariableType>();
+        if (!typeVar)
+          return Action::SkipChildren;
+
+        // If the base is an opened existential type, let's see whether
+        // erase would produce an existential in this case and if so,
+        // we need to check whether any requirements are going to be lost
+        // in process.
+
+        auto existentialType = GetExistentialType(typeVar);
+        if (!existentialType)
+          return Action::SkipChildren;
+
+        auto erasedMemberTy = typeEraseOpenedExistentialReference(
+            Type(member), *existentialType, typeVar, TypePosition::Covariant);
+
+        // If result is an existential type and the base has `where` clauses
+        // associated with its associated types, the call needs a coercion.
+        if (erasedMemberTy->isExistentialType() &&
+            hasConstrainedAssociatedTypes(member)) {
+          RequiresCoercion = true;
+          return Action::Stop;
+        }
+
+        return Action::SkipChildren;
+      }
+
+      // The case where there is a direct access to opened existential type
+      // e.g. `$T` or `[$T]`.
+      if (auto *typeVar = componentTy->getAs<TypeVariableType>()) {
+        if (auto existentialType = GetExistentialType(typeVar)) {
+          RequiresCoercion |= hasConstrainedAssociatedTypes(
+              (*existentialType)->getExistentialLayout());
+          return RequiresCoercion ? Action::Stop : Action::SkipChildren;
+        }
+      }
+
+      return Action::Continue;
+    }
+
+  private:
+    bool hasConstrainedAssociatedTypes(DependentMemberType *member) {
+      auto *assocType = member->getAssocType();
+
+      // If this member has associated type requirements, we are done.
+      if (assocType->getTrailingWhereClause())
+        return true;
+
+      if (auto *DMT = member->getBase()->getAs<DependentMemberType>())
+        return hasConstrainedAssociatedTypes(DMT);
+
+      return false;
+    }
+
+    static Type getBaseTypeOfDependentMemberChain(DependentMemberType *member) {
+      if (!member->getBase())
+        return member;
+
+      auto base = member->getBase();
+
+      if (auto *DMT = base->getAs<DependentMemberType>())
+        return getBaseTypeOfDependentMemberChain(DMT);
+
+      return base;
+    }
+
+    static bool hasConstrainedAssociatedTypes(ExistentialLayout layout) {
+      for (auto *protocol : layout.getProtocols()) {
+        if (llvm::any_of(protocol->getAssociatedTypeMembers(),
+                         [](const auto *assocTypeDecl) {
+                           return bool(assocTypeDecl->getTrailingWhereClause());
```

Comment:
Same concern as above: `getTrailingWhereClause` is more syntactic than semantic.